Raven db 8218

aviv86 opened the pull request.

aviv86 committed:
RavenDB-8218 : JavascriptConversionExtensions : refactoring + add support for DateTime properties (DateTime.Today.Year, user.Birthday.Month, etc)

aviv86 committed:
RavenDB-8218 : document query : allow to add a 'from alias' to query

aviv86 committed:
RavenDB-8218 : add ConvertJsTimeToTimeSpanString() function to ScriptRunner

aviv86 committed:
RavenDB-8218 : Allow users to define custom functions to send to the server using Linq queries + tests

ayende commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -1520,7 +1526,15 @@ private void VisitSelect(Expression operand)
-                            continue;
+                        {
+                            // lambda 2 js
+                            TranslateSelectBodyToJs(newExpression);
```
And what if this is nested?

ayende commented on src/Raven.Client/Documents/Session/IDocumentQuery.cs:
```diff
@@ -115,7 +115,8 @@ public interface IDocumentQuery<T> :
         /// <param name="projections">Array of field projections.</param>
-        IDocumentQuery<TProjection> SelectFields<TProjection>(string[] fields, string[] projections);
+        /// <param name="fromAlias">From token alias.</param>
+        IDocumentQuery<TProjection> SelectFields<TProjection>(string[] fields, string[] projections, string fromAlias = null);
```
Why do we need this `fromAlias` here?

ayende commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -199,87 +225,109 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                {
+                    context.PreventDefault();
+                    var writer = context.GetWriter();
+                    var val = nodeAsConst.Value.ToString().ToLower();
```
Use static values here.

ayende commented on src/Raven.Server/Documents/Patch/ScriptRunner.cs:
```diff
@@ -383,6 +386,22 @@ private JsValue LoadDocument(JsValue self, JsValue[] args)
+                var secs = (ticks /1000) % 60;
+                var ms = ticks % 1000;
+
+                return new JsValue($"{days}.{hours}:{mins}:{secs}.{ms}");
```
Use a timespan here instead of string.

aviv86 committed:
Merge branch 'v4.0' of https://github.com/ravendb/ravendb into RavenDB-8218

aviv86 committed:
RavenDB-8790 : fix bug in LinqMethodsSupport (instead of writing 'this', visit the parameter expression and let ReplaceThisParameter decide if to replace or not)

aviv86 committed:
RavenDB-8218 : use QueryData object in SelectFields()

aviv86 committed:
RavenDB-8218 :
 - BooleanSupport : Use static values
 - ConvertJsTimeToTimeSpanString() : use a timespan instead of string
 - QueriesWithCustomFunctions: add a test for nested selects, test with 'where'

arekpalinski commented on src/Raven.Client/Documents/Queries/QueryData.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public class QueryData
+    {
+        public string[] Fileds { get; set; }
```
typo: Fileds -> Fields

arekpalinski commented on src/Raven.Client/Documents/Queries/QueryData.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Raven.Client.Documents.Queries
+{
+    public class QueryData
```
I wonder if `QueryData` name isn't too wide for this class. Maybe `QuerySelectData`, `QuerySelection`

arekpalinski commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -199,87 +226,108 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                                writer.Write(memberExpression.Member.Name == "UtcNow" ? ".getUTCFullYear()" : ".getFullYear()");
+                                break;
+                            case "Month":
+                                writer.Write(memberExpression.Member.Name == "UtcNow" ? ".getUTCMonth()+1" : ".getMonth()+1");
```
`.getUTCMonth()+1` that's is tricky :)
